src/replication/primary_replication_manager.cpp

```diff
@@ -0,0 +1,155 @@
+  REPLICATION_LOG_TRACE(fmt::format("[SEND] Preparing ReplicateBatchOfRecords."));
+  NOISEPAGE_ASSERT(policy != transaction::ReplicationPolicy::DISABLE, "Replication is disabled, so why are we here?");
+
+  if (policy == transaction::ReplicationPolicy::ASYNC) {
+    // In asynchronous replication, just invoke the commit callbacks immediately.
+    for (const auto &cb : commit_callbacks) {
+      cb.fn_(cb.arg_);
+    }
+  } else {
+    // Copy the commit callbacks into our local list. The callbacks are invoked when the replicas notify the primary
+    // that the replicas have applied their corresponding transactions.
+    {
+      std::unique_lock lock(callbacks_mutex_);
+      // If there are currently no callbacks, execute everything that won't be sent over to the replica.
+      bool was_empty = txn_callbacks_.empty();
+      txn_callbacks_.emplace(commit_callbacks);
+      if (was_empty) {
+        ProcessTxnCallbacks();
+      }
+    }
+  }
+
+  if (records_batch != nullptr) {
+    // Send the batch of records to all replicas.
+    ReplicationMessageMetadata metadata(GetNextMessageId());
+    RecordsBatchMsg msg(metadata, GetNextBatchId(), records_batch);
+    REPLICATION_LOG_TRACE(fmt::format("[SEND] BATCH {}", msg.GetBatchId()));
+
+    messenger::callback_id_t destination_cb =
+        messenger::Messenger::GetBuiltinCallback(messenger::Messenger::BuiltinCallback::NOOP);
+    const msg_id_t msg_id = msg.GetMessageId();
+    const std::string msg_string = msg.ToJson().dump();
+    for (const auto &replica : replicas_) {
+      Send(replica.first, msg_id, msg_string, messenger::CallbackFns::Noop, destination_cb);
+    }
+
+    NOISEPAGE_ASSERT(newest_buffer_txn >= newest_txn_sent_,
+                     "The assumption is that transactions are monotonically increasing.");
+    newest_txn_sent_ = newest_buffer_txn;
+
+    // Return the buffered log writer to the pool if necessary.
+    if (records_batch->MarkSerialized()) {
+      empty_buffer_queue_->Enqueue(records_batch);
+    }
+  }
+}
+
+void PrimaryReplicationManager::NotifyReplicasOfOAT(transaction::timestamp_t oldest_active_txn) {
+  ReplicationMessageMetadata metadata(GetNextMessageId());
+  NotifyOATMsg msg(metadata, last_sent_batch_id_, oldest_active_txn);
+  REPLICATION_LOG_TRACE(fmt::format("[SEND] BATCH {} OAT {}", msg.GetBatchId(), msg.GetOldestActiveTxn()));
+
+  messenger::callback_id_t destination_cb =
+      messenger::Messenger::GetBuiltinCallback(messenger::Messenger::BuiltinCallback::NOOP);
+
+  const msg_id_t msg_id = msg.GetMessageId();
+  const std::string msg_string = msg.ToJson().dump();
+  for (const auto &replica : replicas_) {
+    Send(replica.first, msg_id, msg_string, messenger::CallbackFns::Noop, destination_cb);
+  }
+}
+
+record_batch_id_t PrimaryReplicationManager::GetNextBatchId() {
+  record_batch_id_t next_batch_id = next_batch_id_++;
+  if (next_batch_id_.load() == INVALID_RECORD_BATCH_ID) {
```

Comment:
I don't think this `if` block is atomic, if multiple threads access it at the same time you might skip over a batch id. I think that shouldn't be a problem since I'm pretty sure only one thread is ever accessing this at a time. Though if that's the case it might not make sense to have `next_batch_id` be atomic in the first place. Also, I'm not sure if it matters if we skip over batch ids.

Reply:
1. yes, only one thread is accessing it `LogSerializerTask` invoking `ReplicateBatchOfRecords`
2. you're right, the atomic is unnecessary, made it a regular non-atomic thing
3. just clarification, it is very important that batch IDs are not skipped. The replication log provider on the receiver will hold the batches until the "next" batch ID is seen, and if it is skipped, it won't make progress